### PR TITLE
Java singleton design pattern

### LINKED ISSUE #3407: Java Singleton Design Pattern
### Proposed title of article

Java Singleton Design Pattern

### Introduction paragraph (2-3 paragraphs):

Singleton is a creational design pattern that assures that only one object of its kind exists and gives other code a single point of access to it. Singletons have nearly the same advantages and disadvantages as global variables. They are quite useful, yet they violate the modularity of our code.

In this article, the reader will learn about the singleton design pattern. Understand early Instantiation of Singleton Pattern and lazy Instantiation of Singleton Pattern.

### Key takeaways: 

 In this article we will cover the following;

1.  Singleton Design Pattern
2. Eager loading singleton object
3. Lazy loading singleton object
4. Synchronized with Thread in Lazy Loading
5. Using Thread with Lazy Loading (Double Check Locking Technique)
6. Enum with Singleton Pattern

### References: 
 N/A

### PR DESCRIPTION
**Pull request by carowanjiku**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #3407 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/java-singleton-design-pattern/index.md [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        });
+
+        r1.start();
+        r2.start();
+    }
+}
+
+```
+
+In the example, we find a remedy for this issue of synchronous code overhead. This function's `GetInstance` is not synchronized, but the component that creates the instance is, thus only a few threads must wait the first time.
```
Comment (Dawe-7) [inline content/articles/java-singleton-design-pattern/index.md:299]:
```suggestion
In the example, we find a remedy for this issue of synchronous code overhead. This function's `GetInstance` is not synchronized, but the component that creates the instance is. Thus only a few threads must wait the first time.
```

File: content/articles/java-singleton-design-pattern/index.md [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+### Enum with Singleton Pattern
+We have another way of creating a singleton pattern that overcomes the disadvantages of all four above implementations. From Java 1.5, there is just another technique for creating a singleton design pattern, which is thread-safe and uses fewer resources. This approach can only work if one uses Java 1.5 or higher.
```
Comment (Dawe-7) [inline content/articles/java-singleton-design-pattern/index.md:302]:
```suggestion
This is a way of creating a singleton pattern that overcomes the disadvantages of all the implementations that we looked at. From Java 1.5, there is only one other technique for creating a singleton design pattern which is thread-safe and uses fewer resources. This approach can only work if one uses Java 1.5 or higher.
```

File: content/articles/java-singleton-design-pattern/index.md [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+
+Here is the syntax of an enum with a singleton pattern.
```
Comment (Dawe-7) [inline content/articles/java-singleton-design-pattern/index.md:304]:
```suggestion
Here is the syntax of an enum with a singleton pattern:
```

File: content/articles/java-singleton-design-pattern/index.md [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+This will be the result.
+
+```bash
+15
+25
+
+```
+
+There is an idea of decoding, despite the, we are dealing with Double-Check Locking (DCL). The readObject method will return a new object even if the class is a singleton.
```
Comment (Dawe-7) [inline content/articles/java-singleton-design-pattern/index.md:350]: The first sentence here does not make sense.

File: content/articles/java-singleton-design-pattern/index.md [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+
+When we are working with enums, there is a method called readResolve that will not generate a new object and instead uses the existing one.
```
Comment (Dawe-7) [inline content/articles/java-singleton-design-pattern/index.md:352]:
```suggestion
When we are working with enums, there is a method called `readResolve`  that will not generate a new object and instead uses the existing one.
```

**Review by Dawe-7 — CHANGES_REQUESTED** [review]
@carowanjiku, nice article. I've left some suggestions that i think will improve the quality of your article. Most of the suggestions are as a result of long sentences in your article that are hard to read. I found a few sentences that are hard to understand and i've left a comment with the suggestion. Kindly check on the readability of your article by going through your article again after implementing the suggestions.

> Make sure to pull my changes before implementing the suggestions

Ping me when you are done.

**Comment by hectorkambow** [comment]
@carowanjiku 

Good afternoon and thank you for submitting your topic to the EngEd program. 
If it seems this PR is still quite a bit aways from being completed. 
As a rule of thumb - please only submit articles (pull requests) with a complete and fully polished article - ready to be published. 
All revisions and edits should be completed on your own forked repo (as to not take up room in the queue). 
We will be closing this and please re-submit when it is ready to be published as a new PR (and you can add the duplicate label) and ping the reviewer and myself. 

cc @Dawe-7

**Pull request closed without merging**